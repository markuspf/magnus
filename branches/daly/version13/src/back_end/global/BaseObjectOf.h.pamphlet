\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/BaseObjectOf.h}
\author{Stephane Collart}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{include}
<<include>>=
#include "PureRep.h"
#include "ObjectOf.h"

@
\section{template<class Rep> class BaseObjectOf}
<<template<class Rep> class BaseObjectOf>>=
template<class Rep> class BaseObjectOf : private ObjectOf<Rep>
{
// This assumes that Rep is derived from PureRep
public:
  // copy constructor provided by compiler
  // destructor provided by compiler
  // assignment operator provided by compiler

protected:
  // Inherited from base; adjust access:
  ObjectOf<Rep>::look;
  ObjectOf<Rep>::change;
  ObjectOf<Rep>::enhance;
  ObjectOf<Rep>::acquireRep;
  BaseObjectOf( Rep* newrep ) : ObjectOf<Rep>( newrep) { }

private:
	// Dummy Forcing Member:
	void force_derivation( ) { const PureRep* rc = look(); }
	// With this member PureRep is forced to be 
        // an accessible base class of Rep
};

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/BaseObjectOf.h
#ifndef _BASE_OBJECT_OF_H_
#define _BASE_OBJECT_OF_H_

<<license>>
<<include>>
<<template<class Rep> class BaseObjectOf>>
#endif

@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
