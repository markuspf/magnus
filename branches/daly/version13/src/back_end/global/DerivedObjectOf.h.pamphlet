\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/DerivedObjectOf.h}
\author{Stephane Collart}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{class DerivedObjectOf>
<<class DerivedObjectOf>>=
template <class Base, class Rep> class DerivedObjectOf : public Base
{

// This class assumes Base is itself an ObjectOf kind of class, i.e.
// 1) Base has members look(), enhance(), change() returning some BaseRep* type
// 2) Base has a constructor Base( BaseRep* );
// 3) BaseRep is directly or indirectly derived from PureRep
// 4) Rep is derived from BaseRep

public:
	// copy constructor provided by compiler
	// destructor provided by compiler
	// assignment operator provided by compiler

protected:
	// Shadowed representation accessors:
	const Rep* look( ) const { return (Rep*)Base::look(); }
	Rep* enhance( ) const { return (Rep*)Base::enhance(); }
	Rep* change( ) { return (Rep*)Base::change(); }

	DerivedObjectOf( Rep* newrep ) : Base( newrep ) { }
	// for base class initialisation

	DerivedObjectOf( const Base& b ) : Base( b ) { }
	// for base class intialisation when cast conversion
        // to derived is desired 

};

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/DerivedObjectOf.h
#ifndef _DERIVED_OBJECT_OF_H_
#define _DERIVED_OBJECT_OF_H_

<<license>>
<<class DerivedObjectOf>>
#endif
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
