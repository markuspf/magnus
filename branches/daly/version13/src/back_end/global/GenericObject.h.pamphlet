\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/GenericObject.h}
\author{Stephane Collart}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{include}
<<include>>=
#include "PureRep.h"
#include "BaseObjectOf.h"
#include "Type.h"

@
\section{struct GenericRep}
a common root class for all object-level (see PureObject.h) classes
with a coherent type scheme.

See PureObject

In addition, this class provides explicit type query handles; classes
derived from this class will normally derive *publicly*, to permit
assignment by non-friend
<<struct GenericRep>>=
struct GenericRep : public PureRep
{
	// default constructor provided by compiler
	// copy constructor provided by compiler
	// assignment operator disabled

	static const Type theGenericObjectType;
	// every representation class derived from this 
        // should carry a similarly defined type field

	static Type type() { return theGenericObjectType; }
	// every representation class derived from this 
        // should carry a similarly defined type member

	virtual Type actualType() const { return type(); }
	// this member serves as the root of all type querying members in this
	// type scheme
	// every representation class derived from this should override this
	// member similarly to return the actual type of the class

	PureRep* clone( ) const { return new GenericRep(*this); }

};

@
\section{class GenericObject}
<<class GenericObject>>=
class GenericObject : protected BaseObjectOf<GenericRep>
{
public:
	GenericObject( ) : BaseObjectOf<GenericRep>( new GenericRep ) { }

	// copy constructor provided by compiler
	// destructor provided by compiler
	// both of the preceding ref-counting compatible
	// assignment operator provided by compiler
	// ref-counting compatible

	static Type type() { return GenericRep::type(); }
	// this `type querying' member is provided solely as an example, 
        // since this class is pseudo-abstract (ie. its representation 
        // class is abstract) and no actual objects of this type will 
        // ever be created

	Type actualType() const { return look()->actualType(); }

protected:
	// look(), change() and enhance() inherited

	// Special wrapping constructor to wrap new representations (returned
	// by eg. delegated methods) and for base initialisation by derived
	// classes:

	GenericObject( GenericRep* newrep ) : BaseObjectOf<GenericRep>(newrep) { }

};

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/GenericObject.h
#ifndef _GENERIC_OBJECT_H_
#define _GENERIC_OBJECT_H_


<<include>>
<<license>>
<<struct GenericRep>>
<<class GenericObject>>
#endif
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
