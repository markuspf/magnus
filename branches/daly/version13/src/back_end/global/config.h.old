// Copyright (C) 1994 The New York Group Theory Cooperative
// See magnus/doc/COPYRIGHT for the full notice.

// Contents: Configuration variables set up by the config script
//
// Principal Author: Jon Lennox
//
// Status: Useable.
//
// Revision History:
//

#ifndef _CONFIG_H_
#define _CONFIG_H_

#define MAGNUS_HOME "/usr/bormotov/magnus"
// We keep this compiled-in value to ease debugging.
// Use MagnusHome::magnusHome() in back_end/general/ for
// the Magnus root directory.

#define LINUX

#define MAGNUS_TMP "/tmp"

#endif /* _CONFIG_H_ */
