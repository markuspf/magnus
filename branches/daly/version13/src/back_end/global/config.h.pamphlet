\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/config.h}
\author{Jon Lennox}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/config.h
#ifndef _CONFIG_H_
#define _CONFIG_H_

<<license>>

#define MAGNUS_HOME "/usr/local/magnus"
// We keep this compiled-in value to ease debugging.
// Use MagnusHome::magnusHome() in back_end/general/ for
// the Magnus root directory.

#define LINUX

#define MAGNUS_TMP "/tmp"

#endif /* _CONFIG_H_ */
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
