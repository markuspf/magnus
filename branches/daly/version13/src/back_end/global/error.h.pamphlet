\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/error.h}
\author{Roger Needham}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{error}
<<error>>=
#ifdef __GNUG__
//@jml This tells g++ that error does not return.
inline volatile void error(const char *msg) {
#else
inline void error(const char* msg) {
#endif

  cout << "fatal_error_ {" << msg << '}' << endl << flush;
  // Try to shove the message directly out to the front end
  cerr << endl << "Fatal error: " << msg << endl;
  while ( 1 );
  // So we can get a stack backtrace when debugging
}

@
\section{warn}
<<warn>>=
inline void warn(const char* msg) {
  cerr << endl << "Warning: " << msg << endl;
}

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/error.h
#ifndef _ERROR_H_
#define _ERROR_H_

<<license>>
<<error>>
<<warn>>
#endif
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
