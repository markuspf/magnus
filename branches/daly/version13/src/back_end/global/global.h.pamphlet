\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/global.h}
\author{Roger Needham}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{include}
<<include>>=
#include <stream.h>
#include <iostream.h>
#include <fstream.h>
#include <strstream.h>
//@rn #include <string.h>
#include <ctype.h>
#include <stdlib.h>
#include <math.h>

#include "error.h"
#include "IPC.h"
#include "Trichotomy.h"

@
\section{global.h}
Contents: Global declarations to be included by all other header files.
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/global.h
#ifndef _GLOBAL_H_
#define _GLOBAL_H_

<<include>>
<<license>>

//#define NULL 0

#define SAFETY 2

#define TRUE  1
#define FALSE 0

typedef int Bool;

#define INFINITE 0


//------ global function prototypes, variables, inline functions ------------//

inline int min(int i, int j) { return ((i<j) ? i : j); }
inline int max(int i, int j) { return ((i>j) ? i : j); }
inline int abs(int i) { return ((i < 0) ? -i : i); }
inline int sign(int i) { return (i == 0) ? 0 : ( (i > 0) ? 1 : -1 ); }
#endif
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
