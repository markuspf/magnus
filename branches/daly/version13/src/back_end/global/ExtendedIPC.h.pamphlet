\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/ExtendedIPC.h}
\author{Dmitry Bormotov}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{include}
<<include>>=
#include "global.h"
#include "Integer.h"
#include "Rational.h"

@
\section{Output Operators}
<<Output Operators>>=
inline ostream& operator < ( ostream& ostr, const Integer& n )
{
  ostr << n << ' ';
  return ostr;
}

inline ostream& operator < ( ostream& ostr, const Rational& r )
{
  ostr << r << ' ';
  return ostr;
}

@
\section{Input Operators}
<<Input Operators>>=
inline istream& operator > ( istream& istr, Integer& n )
{
  istr >> n;
  return istr;
}

inline istream& operator > ( istream& istr, Rational& r )
{
  istr >> r;
  return istr;
}

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/ExtendedIPC.h
#ifndef _EXTENDED_IPC_H_
#define _EXTENDED_IPC_H_

<<include>>
<<license>>
<<Output Operators>>
<<Input Operators>>
#endif

@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
