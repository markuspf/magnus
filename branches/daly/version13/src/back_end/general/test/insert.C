// Copyright (C) 1994 The New York Group Theory Cooperative
// See magnus/doc/COPYRIGHT for the full notice.

// Contents: Informal test.
//
// Principal Author: Dmitry Bormotov
//
// Status:
//
// Revision History:
//


#include <stdio.h>
#include <signal.h>
#include <sys/time.h>
#include <unistd.h>
#include <fstream.h>
#include "global.h"

double diffTime(timeval t2, timeval t1)
{
  double result;
  long s = t2.tv_sec - t1.tv_sec;
  result = (t2.tv_usec - t1.tv_usec) * 0.000001 + s;
  //if( t2.tv_usec < t1.tv_usec ) --result;
  return result;
}

void insert( int* y, int a, int j )
{
  int i, key = y[j];
  i = j - 1;

  while( (i >= a) && (y[i] > key) ) {
    y[i+1] = y[i];
    --i;
  }
  y[i+1] = key;
}

int main ( )
{
  timeval t1,t2;
  gettimeofday(&t1, NULL);
  int maxSize = 10000;
  int x[maxSize];

  for( int z = 0; z < 1; ++z ) {
    
    int size = 10000;
     for( int i = 0; i < size; ++i )
      x[i] = size-i-1;
    int a = 0, b = size - 1;
    
    int j = a+1;
    while( j <= b ) {
      insert(x,a,j);
      ++j;
    }
    /*
    if( z == 0 ) {
      for( int i = 0; i < size; ++i )
	cout << " " << x[i];
      cout << endl;
    }
    */
  }

  gettimeofday(&t2, NULL);
  double dif = diffTime(t2,t1);

  cout << "Time: " << dif << endl;
}

