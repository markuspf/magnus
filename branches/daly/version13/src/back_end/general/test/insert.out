Test1 ( 100000 iterations, n = 10,20,30,40,50, the worst case )

./insert (linear search)

0 1 2 3 4 5 6 7 8 9
Time: 0.375824

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19
Time: 1.21795

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29
Time: 2.46386

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39
Time: 4.13788

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49
Time: 6.32309


./insert2 (binary search)

0 1 2 3 4 5 6 7 8 9
Time: 0.575766

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19
Time: 1.60481

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29
Time: 3.01182

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39
Time: 4.87398

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49
Time: 7.11453



Test2 ( 1 iteration, n = 10000, the worst case )

./insert (linear search)
Time: 2.37714

./insert2 (binary search)
Time: 1.88738